### PR TITLE
Add __str__ and _repr_html_ to Variant

### LINKED ISSUE #2367: Display methods for Variant
When we print out a Variant we get ``<tskit.genotypes.Variant object at 0x7f58b0483430>``

We should add a str and repr_html to fix this.

Did this change in 0.5.0? Was Variant a dataclass before, and therefore inherited display methods?

### PR DESCRIPTION
**Pull request by szhan**
Fixes #2367

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Cool! Can you paste in what these look like please so we can have a look?

File: python/tskit/genotypes.py [hunk 1]
```diff
@@ -222,6 +223,27 @@ def copy(self) -> Variant:
...
 
+    def __str__(self) -> str:
+        """
+        Return a plain text summary of the contents of a variant.
+        """
+        variant_rows = [
+            ["Site id", str(self.site.id)],
+            ["Samples", str(len(self.samples))],
+            ["Genotypes", str(len(self.genotypes))],
+            ["Alleles", str(self.num_alleles)],
```
Comment (benjeffery) [inline python/tskit/genotypes.py:234]: Would be nice to give the allele counts here.

File: python/tskit/util.py [hunk 2]
```diff
@@ -542,6 +542,45 @@ def tree_html(tree):
...
+                summary {{cursor: pointer; outline: 0; display: list-item;}}
+              </style>
+              <div class="tskit-table-set">
+                <div class="tskit-table-set-table">
+                  <table class="tskit-table">
+                    <thead>
+                      <tr>
+                        <th style="padding:0;line-height:21px;">
+                          <img style="height: 32px;display: inline-block;padding: 3px 5px 3px 0;" src="https://raw.githubusercontent.com/tskit-dev/administrative/main/tskit_logo.svg"/>
+                          <a target="_blank" href="https://tskit.dev/tskit/docs/latest/python-api.html#the-tree-class"> Tree </a>
```
Comment (benjeffery) [inline python/tskit/util.py:565]: Would be better to link to the documentation for `Variant`.

File: python/tskit/util.py [hunk 3]
```diff
@@ -542,6 +542,45 @@ def tree_html(tree):
...
+                        </th>
+                      </tr>
+                    </thead>
+                    <tbody>
+                      <tr><td>Site Id</td><td>{variant.site.id}</td></tr>
+                      <tr><td>Sample Nodes</td><td>{len(variant.samples)}</td></tr>
+                      <tr><td>Genotypes</td><td>{len(variant.genotypes)}</td></tr>
+                      <tr><td>Alleles</td><td>{variant.num_alleles}</td></tr>
```
Comment (benjeffery) [inline python/tskit/util.py:573]: Again the allele counts would be good.

**Review by benjeffery — COMMENTED** [review]
Good start, needs some more detail like the site position and the allele counts.

File: python/tskit/genotypes.py (hunk 1 above)
Comment (hyanwong) [reply python/tskit/genotypes.py:234]: Would it be nice to implement https://github.com/tskit-dev/tskit/issues/2390 to help with this?

**Comment by benjeffery** [comment]
@szhan Let me know if you want some pointers here.

**Comment by szhan** [comment]
Example output of [`__str__()`.
Screen Shot 2022-07-20 at 5 27 42 PM

**Comment by szhan** [comment]
Example output of `_repr_html_()`.
Screen Shot 2022-07-20 at 5 59 01 PM

File: python/tskit/util.py (hunk 2 above)
Comment (szhan) [reply python/tskit/util.py:565]: The link now points to `Variant` in the doc.

**Comment by szhan** [comment]
The HTML text has really long lines, so it is failing the flake8 test. What should we do?

**Comment by szhan** [comment]
I'm not sure why I can't get `re.match` to work... I'm expecting `re.match` to return a `Match` object not `None`, which is what I'm getting with the code snippet below...

```
import tests.tsutil as tsutil
import re

ts = tsutil.all_fields_ts()
v = next(ts.variants())
str(v)

m = re.match(
    r"allele",
    str(v),
)
print(m)
```

**Comment by szhan** [comment]
> Cool! Can you paste in what these look like please so we can have a look?

I've attached a couple of images showing some examples.

**Comment by szhan** [comment]
Should we also show the ancestral allele?

**Comment by jeromekelleher** [comment]
> The HTML text has really long lines, so it is failing the flake8 test. What should we do?

Looks like you can split some of the longest easily enough (no reason to have lines > 100), but there's also a noqa you can put at the end to tell flake8 to ignore it (see further down in the same function).

**Comment by jeromekelleher** [comment]
> Should we also show the ancestral allele?

Let's show them in the order they're returned by ``variant.alleles``. You can do something like
```python
counts = self.counts()
freqs = self.frequencies()
for allele in self.alleles:
       print(f"Allele {allele}: {count[allele]} ({freqs[allele] * 100: .2g}%)")
```

**Comment by szhan** [comment]
Here is what the output looks like.

`__str__()`
Screen Shot 2022-07-21 at 1 04 50 PM

`_repr_html_()`
Screen Shot 2022-07-21 at 12 56 43 PM

**Comment by hyanwong** [comment]
Looking good, thanks @szhan 

1. Might it be worth saying "Number of alleles" and "Number of samples" rather than just "allele" and "samples"?
2. Do we need the length of genotypes in there (isn't that always the number of samples)?
3. Should we put quotes around the allele string, in case any of the alleles are the empty string (e.g. for an indel). So `Allele 'A'` rather than `Allele A`?
4. Instead of Allele None should we say e.g. "Allele missing"?

**Comment by hyanwong** [comment]
> 1. Might it be worth saying "Number of alleles" and "Number of samples" rather than just "allele" and "samples"?

Actually, I forgot that (a) it might not be all the sample nodes in the ts, if a subset of samples is specified and (b) it is possible to specify non samples to be counted too.

In;m therefore not sure of the best phrasing instead of "samples", but we could say `Number of samples used`?

`Number of

**Comment by hyanwong** [comment]
> Looking good, thanks @szhan
> 
> 1. Might it be worth saying "Number of alleles" and "Number of samples" rather than just "allele" and "samples"? Or `# samples` and `# alleles` if you want a shorter abbreviation.
> 2. Do we need the length of genotypes in there (isn't that always the number of samples)?
> 3. Should we put quotes around the allele string, in case any of the alleles are the empty string (e.g. for an indel). So `Allele 'A'` rather than `Allele A`?
> 4. Instead of Allele None should we say e.g. "Allele missing"?
> 5. "Missing data" might be crueller as "Has missing data"?
> 6. Is isolated as missing meant to be a boolean rather than numeric?

**Comment by szhan** [comment]
> > The HTML text has really long lines, so it is failing the flake8 test. What should we do?
> 
> Looks like you can split some of the longest easily enough (no reason to have lines > 100), but there's also a noqa you can put at the end to tell flake8 to ignore it (see further down in the same function).

There is already `# noqa: B950` at the end of 'variant_html()' in `util.py`.

**Comment by szhan** [comment]
This?
Screen Shot 2022-07-21 at 2 11 16 PM

**Comment by hyanwong** [comment]
> This?
> Screen Shot 2022-07-21 at 2 11 16 PM

Yeah, looks good. Maybe put "Allele missing" instead of `Allele 'None'`?

**Comment by hyanwong** [comment]
Is it clear what the "3" in `3 (2.6%)` means? Should we say "Samples with allele 'A'", or is that too wordy?

Also there's a space between the open bracket and the percentage amount, I think?

**Comment by szhan** [comment]
For the record, I used the following code snippet to produce the HTML representation shown above.

```
from random import random
import tskit
import msprime


ts = msprime.sim_mutations(
    msprime.sim_ancestry(
        5,
        sequence_length=100,
        random_seed=1,
    ),
    rate=0.01,
    random_seed=1,
)

v = next(ts.variants())
v
```

**Comment by szhan** [comment]
The order of alleles in `Variant.alleles` of the first tree of `ts_fixture` is different when I run the code on my laptop versus when run on CircleCI?!

```
def test_variant_str(self, ts_fixture):
        v = next(ts_fixture.variants())
>       assert v.alleles == ("G", "C", "T", "A", None)
E       AssertionError: assert ('C', 'G', 'T', 'A', None) == ('G', 'C', 'T', 'A', None)
E         At index 0 diff: 'C' != 'G'
E         Full diff:
E         - ('G', 'C', 'T', 'A', None)
E         ?       -----
E         + ('C', 'G', 'T', 'A', None)
E         ?  +++++
```

**Comment by jeromekelleher** [comment]
> Should I just change test_variant_str so that the expected allele list matches the list produced when run on CircleCI? But the test may not pass on other machines, e.g. mine.

Yuck. This is why we've been moving away from using simulations as the basis for tests, super annoying dealing with these inconsistencies.

How about creating a very simple tree sequence by hand so you can test? Like
```
tables = tskit.TableCollection(10)
tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
tables.sites.add_row(position=5, ancestral_state="A")
tables.mutations.add_row(site=0, node=0, derived_state="T")
tables.mutations.add_row(site=0, node=1, derived_state="G")
ts = tables.tree_sequence()
```

**Comment by szhan** [comment]
> > Should I just change test_variant_str so that the expected allele list matches the list produced when run on CircleCI? But the test may not pass on other machines, e.g. mine.
> 
> Yuck. This is why we've been moving away from using simulations as the basis for tests, super annoying dealing with these inconsistencies.
> 
> How about creating a very simple tree sequence by hand so you can test? Like
> 
> ```
> tables = tskit.TableCollection(10)
> tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
> tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
> tables.sites.add_row(position=5, ancestral_state="A")
> tables.mutations.add_row(site=0, node=0, derived_state="T")
> tables.mutations.add_row(site=0, node=1, derived_state="G")
> ts = tables.tree_sequence()
> ```

Just introduced `simple_ts` to create a simple dummy ts for testing.

**Comment by szhan** [comment]
All the tests have passed! Updating the change log.

**Comment by hyanwong** [comment]
> Yuck. This is why we've been moving away from using simulations as the basis for tests, super annoying dealing with these inconsistencies.

It's sometime helpful to create a new TS using e.g. `tskit.Tree.generate_balanced(4).tree_sequence`, but that won't have mutations. It is worth having a deterministic routine in `tskit` (or maybe just in the test suite) that adds a site and some mutations to an existing tree sequence in one pass. E.g. 

```
def add_variation(ts, position, ancestral_state, derived_states, mutation_nodes=ts.samples(), mutation_times=None):
    """
    Return a new tree sequence with a new site added at ``position`` with the given ancestral state.
    and for each element $I$ in the list of ``derived states``, add a mutation above mutation_nodes[i] at time
    mutation_times[I] (or TIME_UNKNOWN if not specified)
    """
```

Then making a test case is as simple as

```
ts = tskit.Tree.generate_balanced(4).tree_sequence
ts = tskit.add_variation(ts, 0, "A", ["T", "C"])
```

useful?

**Comment by szhan** [comment]
> > Yuck. This is why we've been moving away from using simulations as the basis for tests, super annoying dealing with these inconsistencies.
> 
> It's sometime helpful to create a new TS using e.g. `tskit.Tree.generate_balanced(4).tree_sequence`, but that won't have mutations. It is worth having a deterministic routine in `tskit` (or maybe just in the test suite) that adds a site and some mutations to an existing tree sequence in one pass? E.g.
> 
> ```
> def add_variation(ts, position, ancestral_state, derived_states, mutation_nodes=ts.samples(), mutation_times=None):
>     """
>     Return a new tree sequence with a new site added at ``position`` with the given ancestral state.
>     and for each element $i$ in the list of ``derived states``, add a mutation above mutation_nodes[i]
>     at time mutation_times[i] (or TIME_UNKNOWN if not specified)
>     """
> ```
> 
> Then making a test case is as simple as
> 
> ```
> ts = tskit.Tree.generate_balanced(4).tree_sequence
> ts = tskit.add_variation(ts, 0, "A", ["T", "C"])
> ```
> 
> useful?

Should we add this in separate PR?

**Comment by jeromekelleher** [comment]
> Should we add this in separate PR?

We should split it off to a separate issue at least.

**Comment by szhan** [comment]
> > Should we add this in separate PR?
> 
> We should split it off to a separate issue at least.

Should we do that before finishing this PR?

**Comment by jeromekelleher** [comment]
Let's not bother with it, it's just a utility for testing. @hyanwong can create an to discuss  as it's his idea.

**Comment by szhan** [comment]
When there are no sample nodes selected to get the genotypes, the outputs look like this.

`__str__()`
Screen Shot 2022-07-23 at 6 59 31 PM

`_repr_html_()`
Screen Shot 2022-07-23 at 6 57 03 PM

Also, it gives this warning:
`WARNING:root:No non-missing samples at this site, frequencies undefined`

**Comment by szhan** [comment]
When trying to call `_str_()` and `_repr_html_()` on an un-decoded `Variant` object, we get this error:

```
ValueError: This variant has not yet been decoded at a specific site,call Variant.decode to set the site.
```

This should be the behaviour we want, right?

**Comment by benjeffery** [comment]
> When trying to call `_str_()` and `_repr_html_()` on an un-decoded `Variant` object, we get this error:
> 
> ```
> ValueError: This variant has not yet been decoded at a specific site,call Variant.decode to set the site.
> ```
> 
> This should be the behaviour we want, right?

I don't think so, it should be a one-line table with "This variant has not yet been decoded at a specific site,call Variant.decode to set the site" as the content IMO.

**Comment by szhan** [comment]
Something like this?

╔═══════════════════════════════════════════════════════════╗
║ This variant has not yet been decoded at a specific site, call Variant.decode to set the site.║
╚═══════════════════════════════════════════════════════════╝

**Comment by jeromekelleher** [comment]
Yes, ``__str__``  etc should always succeed even if the variant hasn't been decoded yet.

**Review by jeromekelleher — APPROVED** [review]
LGTM!

File: python/tests/test_genotypes.py [hunk 4]
```diff
@@ -2100,3 +2102,60 @@ def test_variant_frequencies_no_samples(self, ts_fixture, caplog):
...
+                ╟─+┼─+╢
+                ║Isolated as missing\s*│\s*True║
+                ╚═+╧═+╝
+                """[
+                    1:
+                ]
+            ),
+            str(v),
+        )
+
```
Comment (hyanwong) [inline python/tests/test_genotypes.py:2155]: Can you add in a check for no samples and also when the variant isn't set to a site?

```python
def test_variant_str_no_samples():
    v = next(ts_fixture.variants(samples=[]))
    for allele in v.alleles:
        if allele is not None:
            assert re.search(f"║Samples with allele '{allele}'\s*│\s*0\s*\(nan\%\)║", str(v)) is not None

def test_variant_str_no_site():
   v = tskit.Variant(ts_fixture)
   s = str(v)
   assert len(s.splitlines()) == 3
   assert "This variant has not yet been decoded at a specific site, call Variant.decode to set the site" in s


```

**Review by hyanwong — APPROVED** [review]

**Comment by szhan** [comment]
When calling `__str__()` and `_repr_html_()` on an un-decoded `Variant` object, the outputs look like this.

Screen Shot 2022-07-27 at 3 12 11 PM
Screen Shot 2022-07-27 at 3 11 54 PM

File: python/tests/test_genotypes.py [hunk 5]
```diff
@@ -2100,3 +2102,60 @@ def test_variant_frequencies_no_samples(self, ts_fixture, caplog):
...
+        It has only one tree and one site, whose variant has the alleles
+        ('A', 'T', 'G', None).
+        """
+        tables = tskit.TableCollection(10)
+        tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
+        tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
+        tables.nodes.add_row(flags=tskit.NODE_IS_SAMPLE, time=0)
+        tables.sites.add_row(position=5, ancestral_state="A")
+        tables.mutations.add_row(site=0, node=0, derived_state="T")
+        tables.mutations.add_row(site=0, node=1, derived_state="G")
```
Comment (hyanwong) [inline python/tests/test_genotypes.py:2118]: Can we put one of the derived states to a unicode character, e.g. the classic 💩 ?!

Comment (hyanwong) [reply python/tests/test_genotypes.py:2118]: Oh, and set one of the others to '' (empty string) and other to multiple characters, e.g. `AAC`

Comment (szhan) [reply python/tests/test_genotypes.py:2118]: The poop emoji messes up the alignment of the columns...
Screen Shot 2022-07-27 at 4 23 41 PM

Comment (szhan) [reply python/tests/test_genotypes.py:2117]: It is because the poop emoji is of the same width as the other characters.

Comment (hyanwong) [reply python/tests/test_genotypes.py:2117]: I don't think we care. This is an edge case. But it's worth keeping a weird character in the tests to check that we don't bug out when trying to print, etc.

**Review by hyanwong — APPROVED** [review]
LGTM

File: python/tskit/genotypes.py [hunk 6]
```diff
@@ -294,6 +295,45 @@ def frequencies(self, remove_missing=None) -> dict[str, float]:
             if not (allele is None and remove_missing)
         }
 
+    def __str__(self) -> str:
+        """
+        Return a plain text summary of the contents of a variant.
+        """
+        try:
+            self.site
```
Comment (benjeffery) [inline python/tskit/genotypes.py:303]: Is this line needed? I know the intent is to check the variant is decoded, but that happens a few lines lower anyway.

File: python/tskit/util.py [hunk 7]
```diff
@@ -542,6 +542,98 @@ def tree_html(tree):
...
+        freqs = variant.frequencies()
+
+        return (
+            html_body_head
+            + f"""
+                <tr><td>Site Id</td><td>{site_id}</td></tr>
+                <tr><td>Site Position</td><td>{site_position}</td></tr>
+                <tr><td>Number of Samples</td><td>{num_samples}</td></tr>
+                <tr><td>Number of Alleles</td><td>{num_alleles}</td></tr>
+            """  # noqa: B950
```
Comment (benjeffery) [inline python/tskit/util.py:609]: Do you still need the `noqa` here?

File: python/tskit/util.py [hunk 8]
```diff
@@ -542,6 +542,98 @@ def tree_html(tree):
...
+                <tr><td>Isolated as Missing</td><td>{isolated_as_missing}</td></tr>
+            """  # noqa: B590
+            + html_body_tail
+        )
+    except ValueError as err:
+        return (
+            html_body_head
+            + f"""
+                        <tr><td>Error</td><td>{str(err)}</td></tr>
+            """  # noqa: B950
```
Comment (benjeffery) [inline python/tskit/util.py:632]: This one aswell?

File: python/tskit/util.py [hunk 9]
```diff
@@ -542,6 +542,98 @@ def tree_html(tree):
...
+            )
+            + f"""
+                <tr><td>Has Missing Data</td><td>{has_missing_data}</td></tr>
+                <tr><td>Isolated as Missing</td><td>{isolated_as_missing}</td></tr>
+            """  # noqa: B590
+            + html_body_tail
+        )
+    except ValueError as err:
+        return (
+            html_body_head
```
Comment (benjeffery) [inline python/tskit/util.py:629]: Might as well get this covered.

File: python/tskit/util.py [hunk 10]
```diff
@@ -542,6 +542,98 @@ def tree_html(tree):
...
+                    + " "
+                    + f"({freqs[k] * 100:.2g}%)"
+                    + "</td></tr>"
+                    for k in variant.alleles
+                ]
+            )
+            + f"""
+                <tr><td>Has Missing Data</td><td>{has_missing_data}</td></tr>
+                <tr><td>Isolated as Missing</td><td>{isolated_as_missing}</td></tr>
+            """  # noqa: B590
```
Comment (benjeffery) [inline python/tskit/util.py:624]: And here, I don't think 590 is a valid bugbear code anyway.

**Review by benjeffery — APPROVED** [review]
Awesome! Just a couple of nits.

File: python/tskit/util.py (hunk 9 above)
Comment (szhan) [reply python/tskit/util.py:629]: I've added this test.

```
def test_variant_html_repr_no_site(self):
        tables = tskit.TableCollection(10)
        ts = tables.tree_sequence()
        v = tskit.Variant(ts)
        html = v._repr_html_()
        ElementTree.fromstring(html)
        assert len(html) > 1600
```

But I'm not sure if 1600 is appropriate here. The length of the HTML string is 1633.
